106-self-healing-feature-in-operator.md

```diff
@@ -0,0 +1,295 @@
+* **RebalanceOnAnomalyDetection**: a rebalancing related to a self-healing of cluster.
+
+The FSM states' transitions are the following:
+
+* from **Idle** to:
+  * **RebalanceOnAnomalyDetection**: if the `strimzi.io/anomaly-detected: <anomaly-type>` annotation is applied on the Kafka CR.
+* from **RebalanceOnAnomalyDetection**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is queued and will run if there is no other rebalancing scale down in queue. If a rebalancing scale down is in queue then it will be executed first.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, it will run once the full rebalance is completed
+  * **Idle**: if full rebalance was requested, it was executed and completed successfully or failed.
+
+On each reconciliation, the logic will be like this:
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if the annotation `strimzi.io/anomaly-detected: <anomaly-type>` is applied on the Kafka CR, if applied then the `Full` rebalance would be performed.
+   What happens for each reconciliation depends on the auto-rebalancing FSM as well as the "actual" `KafkaRebalance` custom resource status.
+
+The `KafkaAutoRebalancingReconciler.reconcile()` loads the `Kafka.status.autoRebalance` content:
+
+* `state`: is the FSM state.
+* `lastTransitionTime`: when the transition to that state happened.
+* `modes`: sets the mode as `Full`
+
+Let's see what happens during the auto-rebalancing process when the FSM starts from the **Idle** state and tranistions to **RebalanceOnAnomalyDetection**
+
+#### Idle
+
+This state is set since the beginning when a `Kafka` custom resource is created with the `spec.cruiseControl.autoRebalance` field.
+It is also the end state of a previous successfully completed or failed auto-rebalancing.
+In this state, the operator removes the finalizer and deletes the corresponding "actual" `KafkaRebalance` custom resource.
+
+#### RebalanceOnAnomalyDetection
+
+In this state, an anomaly is detected and the annotation `strimzi.io/anomaly-detected: <anomaly-type>` is applied on `Kafka` CR
+
+Check the current `KafkaRebalance` status:
+
+* if `Ready`, the rebalance was successful.
+  * if there is a queued rebalancing scale down (`Kafka.status.autoRebalance.modes[remove-brokers]` exists), start the rebalancing scale down and transition to **RebalanceOnScaleDown**.
+  * if there is a queued rebalancing scale up (`Kafka.status.autoRebalance.modes[add-brokers]` exists), start the rebalancing scale up and transition to **RebalanceOnScaleUp**.
+  * If no queued rebalancing scale down or scale up, just transition to **Idle**, clean `Kafka.status.autoRebalance.modes`, delete the "actual" `KafkaRebalance` custom resource and remove the annotation .
+* if `PendingProposal`, `ProposalReady` or `Rebalancing`, the rebalancing scale up is still running.
```

Comment:
Full rebelancing due to anomaly not scale up (I guess just a copy/paste mistake here)